### ISSUE TITLE
Castle screen, fully built castle, differences noticed

### ISSUE THREAD

**Issue by LeHerosInconnu** [ihhub/fheroes2#2059]
In fheroes2, there are defects due to the display priority of sprites or/and sprites which are displayed differently.
In fheroes2, some defects have already been corrected compared to the original version of the game, they have not been reported here.

**Sorceress' Castle.**
- Pixels at the bottom and to the left of the Captain's Quarters do not have the same color as in the original game.

In original game:

![Castle screen H2 003](https://user-images.githubusercontent.com/43583994/96756233-13bbc780-13d4-11eb-95c1-353c8b9a1c7b.png)

In fheroes2:

![Castle screen fh2 003](https://user-images.githubusercontent.com/43583994/96756245-19b1a880-13d4-11eb-9963-c324a3358a5a.png)

Open the two images in different tabs in your browser, enlarge the images and switch between the two tabs alternately.
Heroes 2 first, fheroes2 second.

![Castle screen H2 captain coupé 003](https://user-images.githubusercontent.com/43583994/96756263-20402000-13d4-11eb-8e85-945bad2ad8ad.png) ![Castle screen fh2 captain coupé 003](https://user-images.githubusercontent.com/43583994/96756272-23d3a700-13d4-11eb-9288-16e9b593d4a3.png)

**Edit 2020/12/21.**

**Knight's Castle.**

To comply with the creature buildings color code, red for unimproved buildings and blue for upgraded buildings.
Knight castle, inverted the sprites for the Archery Range and the Upg. Archery Range (red and blue).

TWNKDW_1.ICN
0000.bmp
- replaced by 0000.bmp from TWNKUP_1.ICN

TWNKUP_1.ICN
0000.bmp
- replaced by 0000.bmp from TWNKDW_1.ICN

CSTLKNGT.ICN
0020.bmp
- renamed to 0025.bmp

0025.bmp
- renamed to 0020.bmp

**Necromancer's Castle.**

There is a problem with the Laboratory road connection.

In fheroes2:
![Laboratory road flaw fh2 01](https://user-images.githubusercontent.com/43583994/102753862-58c88f80-436c-11eb-83d5-86773b6ce1fe.png)

In original game, corrected:
![Laboratory road corrected H2 01](https://user-images.githubusercontent.com/43583994/102753854-55cd9f00-436c-11eb-9b0a-38e507c33d8a.png)

There is a problem with the Mage Guild road connection.

In fheroes2:
![Mage Guild road flaw fh2 01](https://user-images.githubusercontent.com/43583994/102754107-c2489e00-436c-11eb-84f1-885cff52d98e.png)

In original game:
![Mage Guild H2 01](https://user-images.githubusercontent.com/43583994/102754118-c4aaf800-436c-11eb-8a74-5b7cb8cf7e81.png)

Necromancer castle screen, changes needed to avoid display "flaws" when switching between castles with Mausoleum and upgraded Mausoleum.
The position of the sprites must be adjusted.
Note: Graphic modifications have also been applied, they will be specified in another issue (changes to the half top of the sprite).

TWNNDW_4.ICN
0000.bmp

TWNNUP_4.ICN
0000.bmp

In fheroes2:
![Dwelling Lich position fh2 01](https://user-images.githubusercontent.com/43583994/102754388-371bd800-436d-11eb-8996-7a9be869979c.png)

In fheroes2:
![Dwelling Lich position fh2 02](https://user-images.githubusercontent.com/43583994/102754392-397e3200-436d-11eb-9628-5a4dabc870dc.png)

In original game, corrected:
![Dwelling Lich position corrected H2 02](https://user-images.githubusercontent.com/43583994/102754397-3be08c00-436d-11eb-8262-2e4ba9e1230a.png)

Open the three images in different tabs in your browser, enlarge the images and switch between the three tabs alternately.
fheroes2 unimproved dwelling first, fheroes2 upgraded dwelling second, Heroes 2 corrected position third.

![Dwelling Lich position fh2 coupé 01](https://user-images.githubusercontent.com/43583994/102755032-1e5ff200-436e-11eb-815e-a93a987d4b74.png) ![Dwelling Lich position fh2 coupé 02](https://user-images.githubusercontent.com/43583994/102755033-1f911f00-436e-11eb-99e6-43167285f87a.png) ![Dwelling Lich position corrected H2 coupé 02](https://user-images.githubusercontent.com/43583994/102755031-1dc75b80-436e-11eb-8798-d6f08ba25191.png)

All changes applied (graphics change included).
Heroes 2 original first, Heroes 2 corrected second.
![Original-0000](https://user-images.githubusercontent.com/43583994/102755271-73036d00-436e-11eb-8974-fc28d3285839.png) ![Change-0000](https://user-images.githubusercontent.com/43583994/102755275-739c0380-436e-11eb-9272-070caa581abb.png)

**Comment by ihhub** [comment]
Hi @LeHerosInconnu , I've updated the order of building drawing. I didn't close this issue. Please check it again (once the compilation is done) and let me know if something is missing or simply close the issue if everything is fine. What I fixed:
- Tavern drawing in Barbarian castle
- Captain drawing in Necromancer castle
- Statue drawing in Knight castle
- Library drawing in Wizard castle

Of cause not everything could be fixed as I noticed that the original game has even more issues than our.

**Comment by LeHerosInconnu** [comment]
@ihhub.

> Hi @LeHerosInconnu , I've updated the order of building drawing. I didn't close this issue. Please check it again (once the compilation is done) and let me know if something is missing or simply close the issue if everything is fine. What I fixed:
> 
> * Tavern drawing in Barbarian castle
> * Captain drawing in Necromancer castle
> * Statue drawing in Knight castle
> * Library drawing in Wizard castle
> 
> Of course not everything could be fixed as I noticed that the original game has even more issues than our.

All the changes you have made are working properly.
For other problems, it will certainly be necessary to slightly modify the graphics because a change in priority will not necessarily solve the problem.

I suggest leaving the issue open as a reminder for the other problems.
If it's ok, I'll edit the first post to leave only the problems unfixed.

**Comment by ihhub** [comment]
Yes, please edit the first message. Maybe these pixel related issues because of another incorrect ICN decoding.

**Comment by LeHerosInconnu** [comment]
@ihhub.

> Yes, please edit the first message. Maybe these pixel related issues because of another incorrect ICN decoding.

I edited the first post.

**Comment by ihhub** [comment]
> @ihhub.
> 
> > Yes, please edit the first message. Maybe these pixel related issues because of another incorrect ICN decoding.
> 
> I edited the first post.

Mmm.. I don't see any changes :(

**Comment by LeHerosInconnu** [comment]
@ihhub.

> > @ihhub.
> > > Yes, please edit the first message. Maybe these pixel related issues because of another incorrect ICN decoding.
> > 
> > 
> > I edited the first post.
> 
> Mmm.. I don't see any changes :(

Maybe I hadn't updated the change. :)
It should be fine now.

**Comment by ihhub** [comment]
@LeHerosInconnu well, it happens. No worries :)

**Comment by ihhub** [comment]
Hi @LeHerosInconnu could you please verify castle drawings wit latest commit? I made fixes for Barbarian castle roads.

Regarding different pixels in Knight and Sorceress castles: it might be that we miss some extra logic about image transcoding. I made some changes for images but it looks like it's not the case. So I don't know what's wrong with them.

**Comment by LeHerosInconnu** [comment]
Hello @ihhub,

> Hi @LeHerosInconnu could you please verify castle drawings with latest commit? I made fixes for Barbarian castle roads.
> 
> Regarding different pixels in Knight and Sorceress castles: it might be that we miss some extra logic about image transcoding. I made some changes for images but it looks like it's not the case. So I don't know what's wrong with them.

Your corrections for the routes of the barbarian castles are working correctly.
It's perfect. :)
Did you modify the sprites "in the program", by cutting part of the sprites for example ?

In these castles screens, there are also other flaws that have not been corrected in fheroes2, do you want me to report them in this thread, or it can wait ?

I've modified the first post accordingly.

**Comment by ihhub** [comment]
Hi @LeHerosInconnu , I didn't modify any sprites but made a correct order of drawings.

If you spot some issues with sprites like incorrect pixels or colors, for example Calavry sprite which we fixed yesterday, please create a thread for this. We might able to fix it within an application without using modified assets of the game.

**Comment by LeHerosInconnu** [comment]
@ihhub.

> Hi @LeHerosInconnu , I didn't modify any sprites but made a correct order of drawings.
> 
> If you spot some issues with sprites like incorrect pixels or colors, for example Calavry sprite which we fixed yesterday, please create a thread for this. We might able to fix it within an application without using modified assets of the game.

Okay, I will add the other defects of the castles in this thread and create another thread for the other defects.

**Comment by LeHerosInconnu** [comment]
First post updated with other flaws for Knight's Castle and Necromancer's Castle.

**Comment by ihhub** [comment]
Hi @LeHerosInconnu , could you please update the first post in regards to Knight castle since we fixed the issue with the Farm?

**Comment by LeHerosInconnu** [comment]
Hello @ihhub,

> Hi @LeHerosInconnu , could you please update the first post in regards to Knight castle since we fixed the issue with the Farm?

First post updated.

**Comment by Eldorias-Piano-Covers** [comment]
Here is an animated gif concerning the Mosoleum pixel offset I was about to post :
![A](https://user-images.githubusercontent.com/104630245/166425321-677fd61d-1b17-432a-a321-be8ba671a4ec.gif)

**Comment by ihhub** [comment]
Reclassifying this issue as asset enhancement with target for beyond 1.0.
